src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-    int nlocal_fine_dofs, nlocal_coarse_dofs;
-    mfem::Array<int> local_fine_dofs;
+    DenseMatrix evects;
+    LocalEigenSolver eigs(max_evects_, spect_tol_);
 
-    int* Pvertices_i = new int[nvertices + 1];
-    Pvertices_i[0] = 0;
-    for (unsigned int i = 0; i < nAggs; ++i)
+    for (int agg = 0; agg < num_aggs; ++agg)
     {
-        GetTableRow(Agg_vertex, i, local_fine_dofs);
-        nlocal_coarse_dofs = vertex_target[i].Width();
-        nlocal_fine_dofs = local_fine_dofs.Size();
-        for (int j = 0; j < nlocal_fine_dofs; ++j)
-            Pvertices_i[local_fine_dofs[j] + 1] = nlocal_coarse_dofs;
+        std::vector<int> edge_dofs_ext = GetExtDofs(gt_.agg_ext_edge_, agg);
+        std::vector<int> vertex_dofs_ext = GetExtDofs(gt_.agg_ext_vertex_, agg);
+        std::vector<int> vertex_dofs_local = gt_.agg_vertex_local_.GetIndices(agg);
+
+        if (edge_dofs_ext.size() == 0)
+        {
+            vertex_targets_[agg] = DenseMatrix(1, 1, {1.0});
+            continue;
+        }
+
+        SparseMatrix M_sub = M_ext.GetSubMatrix(edge_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub = D_ext.GetSubMatrix(vertex_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub_T = D_sub.Transpose();
+
+        D_sub_T.InverseScaleRows(M_sub);
+
+        SparseMatrix DMinvDT = D_sub.Mult(D_sub_T);
+
+        eigs.Compute(DMinvDT, evects);
+
+        if (evects.Cols() > 1)
+        {
+            DenseMatrix evects_ortho = evects.GetCol(1, evects.Cols());
+            agg_ext_sigma_[agg] = D_sub_T.Mult(evects_ortho);
+        }
+        else
+        {
+            agg_ext_sigma_[agg].SetSize(D_sub_T.Rows(), 0);
+        }
+
+        DenseMatrix evects_restricted = RestrictLocal(evects, col_marker_,
+                                                      vertex_dofs_ext, vertex_dofs_local);
+
+        VectorView first_vect = evects_restricted.GetColView(0);
+        vertex_targets_[agg] = smoothg::Orthogonalize(evects_restricted, first_vect, 1, max_evects_);
     }
-    for (int i = 0; i < nvertices; ++i)
-        Pvertices_i[i + 1] += Pvertices_i[i];
+}
 
-    int* Pvertices_j = new int[Pvertices_i[nvertices]];
-    double* Pvertices_data = new double[Pvertices_i[nvertices]];
+std::vector<std::vector<DenseMatrix>> GraphCoarsen::CollectSigma(const SparseMatrix& face_edge)
+{
+    SharedEntityComm<DenseMatrix> sec_sigma(gt_.face_true_face_);
+
+    int num_faces = gt_.face_edge_local_.Rows();
 
-    int coarse_vertex_dof_counter = 0;
-    int ptr;
-    for (unsigned int i = 0; i < nAggs; ++i)
+    for (int face = 0; face < num_faces; ++face)
     {
-        nlocal_coarse_dofs = vertex_target[i].Width();
-        GetTableRow(Agg_vertex, i, local_fine_dofs);
-        nlocal_fine_dofs = local_fine_dofs.Size();
-        mfem::DenseMatrix& target_i = vertex_target[i];
-        for (int j = 0; j < nlocal_fine_dofs; ++j)
+        std::vector<int> face_dofs = face_edge.GetIndices(face);
```

Comment:
I suggest we check here if size of `face_dofs` is 1, in which case we can send an empty matrix. Similarly for `CollectD` and `CollectM`. 

This can avoid some unnecessary communication when we have two big aggregates that connect to another one through only one edge.  I added this condition in the original code when I encountered some big graph (it was a YouTube graph) that has some hubs.